service/matching/task_validation.go

```diff
@@ -0,0 +1,167 @@
+	"time"
+
+	commonpb "go.temporal.io/api/common/v1"
+	enumspb "go.temporal.io/api/enums/v1"
+	"go.temporal.io/api/serviceerror"
+
+	"go.temporal.io/server/api/historyservice/v1"
+	persistencespb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/service/worker/scanner/taskqueue"
+)
+
+const (
+	taskReaderOfferTimeout        = 60 * time.Second
+	taskReaderValidationThreshold = 600 * time.Second
+)
+
+type (
+	taskValidator interface {
+		maybeValidate(
+			task *persistencespb.AllocatedTaskInfo,
+			taskType enumspb.TaskQueueType,
+		) bool
+	}
+
+	taskValidationInfo struct {
+		taskID         int64
+		validationTime time.Time
+	}
+
+	taskValidatorImpl struct {
+		historyClient historyservice.HistoryServiceClient
+
+		lastValidatedTaskInfo *taskValidationInfo
+	}
+)
+
+func newTaskValidator(
+	historyClient historyservice.HistoryServiceClient,
+) *taskValidatorImpl {
+	return &taskValidatorImpl{
+		historyClient: historyClient,
+	}
+}
+
+// probabilistically validate whether a task has expired
+// if return false, nil, then task is invalid and should be discarded
+// if return true, nil, then task maybe valid, and should be dispatched
+func (v *taskValidatorImpl) maybeValidate(
+	task *persistencespb.AllocatedTaskInfo,
+	taskType enumspb.TaskQueueType,
+) bool {
+	if taskqueue.IsTaskExpired(task) {
+		return false
+	}
+	if !v.shouldValidate(task) {
+		return true
+	}
+	valid, err := v.isTaskValid(task, taskType)
+	if err != nil {
+		return true
+	}
+	v.lastValidatedTaskInfo = &taskValidationInfo{
+		taskID:         task.TaskId,
+		validationTime: time.Now().UTC(),
+	}
+	return valid
+}
+
+func (v *taskValidatorImpl) shouldValidate(
+	task *persistencespb.AllocatedTaskInfo,
+) bool {
+	if v.lastValidatedTaskInfo == nil || v.lastValidatedTaskInfo.taskID != task.TaskId {
+		// this task has not been validated
+
+		// after timeout attempting to dispatch the task, checking whether the task is still valid
+		if task.Data.CreateTime != nil && time.Now().Sub(*task.Data.CreateTime) > taskReaderValidationThreshold {
+			return true
+		}
+	} else {
+		// this task has been validated before
+		if time.Now().Sub(v.lastValidatedTaskInfo.validationTime) > taskReaderValidationThreshold {
+			return true
+		}
+	}
+	return false
+}
+
+func (v *taskValidatorImpl) isTaskValid(
+	task *persistencespb.AllocatedTaskInfo,
+	taskType enumspb.TaskQueueType,
+) (bool, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 4*time.Second)
+	defer cancel()
+
+	namespaceID := task.Data.NamespaceId
+	workflowID := task.Data.WorkflowId
+	runID := task.Data.RunId
+
+	switch taskType {
+	case enumspb.TASK_QUEUE_TYPE_ACTIVITY:
+		resp, err := v.historyClient.IsActivityTaskValid(ctx, &historyservice.IsActivityTaskValidRequest{
+			NamespaceId: namespaceID,
+			Execution: &commonpb.WorkflowExecution{
+				WorkflowId: workflowID,
+				RunId:      runID,
+			},
+			Clock:            task.Data.Clock,
+			ScheduledEventId: task.Data.ScheduledEventId,
+		})
```

Comment:
Seems like we've introduced a new round trip on the critical path for task processing. Have you thought about performance regressions? It seems like the fact that we only perform this check for some tasks will help us, but I'm wondering what percent of tasks that will be, and, if the fact that we repeatedly perform this check could block other tasks.

Reply:
it's not quite on the critical path, these tasks have already made a round trip to the db and back. and it's only if they're 10m old, so time is not quite of the essence anymore.

I think there's a slight optimization we could do, which is to only try to validate after a dispatch timeout, even if the task is > 10m old. currently I think it will try to validate before trying to match the task even if there's a poller waiting, which is a waste since Record*TaskStarted will do the same validation.

one quick change could be to keep an attempt count in dispatchBufferedTasks and validate if `attempts % 10 == 2` (i.e. on the 2nd, 12th, 22nd, .. attempts), which should have about the same effect as the current code but skip the first attempt always.

Reply:
>Seems like we've introduced a new round trip on the critical path here. Have you thought about performance regressions?

matching service will not try to make RPC call to history for each task. matching service will only make RPC to history if
1. timeout attempting to dispatch a task (1m timeout)
2. task was created 10m+ ago

>I think there's a slight optimization we could do, which is to only try to validate after a dispatch timeout, even if the task is > 10m old.

Reply:
>I think there's a slight optimization we could do, which is to only try to validate after a dispatch timeout, even if the task is > 10m old. currently I think it will try to validate before trying to match the task even if there's a poller waiting, which is a waste since Record*TaskStarted will do the same validation.

i guess there are few preferences of how to check if a task is still valid
1. check before dispatching
2. check after dispatching timeout

logic before this PR does # 1, e.g. call IsTaskExpired before task dispatching
i guess it is also preferable to unify task expiration check in one place, so this is why expiration check is done in this PR before dispatching.

i understand that ^ optimization can be done to save one RPC call, but this case to me is rare, e.g. task is > 10m old.

let me know if you still want ^ optimization